src/base/process_util.c

```diff
@@ -0,0 +1,223 @@
+      if (hStdIn != INVALID_HANDLE_VALUE)
+	{
+	  SetHandleInformation (hStdIn, HANDLE_FLAG_INHERIT, HANDLE_FLAG_INHERIT);
+	  start_info.dwFlags = STARTF_USESTDHANDLES;
+	  start_info.hStdInput = hStdIn;
+	  inherit_flag = TRUE;
+	}
+    }
+  if (stdout_file)
+    {
+      hStdOut =
+	CreateFile (stdout_file, GENERIC_WRITE, FILE_SHARE_READ, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+      if (hStdOut != INVALID_HANDLE_VALUE)
+	{
+	  SetHandleInformation (hStdOut, HANDLE_FLAG_INHERIT, HANDLE_FLAG_INHERIT);
+	  start_info.dwFlags = STARTF_USESTDHANDLES;
+	  start_info.hStdOutput = hStdOut;
+	  inherit_flag = TRUE;
+	}
+    }
+  if (stderr_file)
+    {
+      hStdErr =
+	CreateFile (stderr_file, GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+		    CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+      if (hStdErr != INVALID_HANDLE_VALUE)
+	{
+	  SetHandleInformation (hStdErr, HANDLE_FLAG_INHERIT, HANDLE_FLAG_INHERIT);
+	  start_info.dwFlags = STARTF_USESTDHANDLES;
+	  start_info.hStdError = hStdErr;
+	  inherit_flag = TRUE;
+	}
+    }
+
+  res =
+    CreateProcess (argv[0], cmd_arg_ptr, NULL, NULL, inherit_flag, CREATE_NO_WINDOW, NULL, NULL, &start_info, &proc_info);
+  free(cmd_arg_ptr);
+
+  if (hStdIn != INVALID_HANDLE_VALUE)
+    {
+      CloseHandle (hStdIn);
+    }
+  if (hStdOut != INVALID_HANDLE_VALUE)
+    {
+      CloseHandle (hStdOut);
+    }
+  if (hStdErr != INVALID_HANDLE_VALUE)
+    {
+      CloseHandle (hStdErr);
+    }
+
+  if (res == FALSE)
+    {
+      return -1;
+    }
+
+  new_pid = proc_info.dwProcessId;
+
+  if (wait_flag)
+    {
+      DWORD status = 0;
+      WaitForSingleObject (proc_info.hProcess, INFINITE);
+      GetExitCodeProcess (proc_info.hProcess, &status);
+      if (exit_status != NULL)
+	*exit_status = status;
+      CloseHandle (proc_info.hProcess);
+      CloseHandle (proc_info.hThread);
+      return 0;
+    }
+  else
+    {
+      CloseHandle (proc_info.hProcess);
+      CloseHandle (proc_info.hThread);
+      return new_pid;
+    }
+}
+#else
+  int pid, rc;
+
+  if (exit_status != NULL)
+    *exit_status = 0;
+
+  if (wait_flag)
+    signal (SIGCHLD, SIG_DFL);
+  else
+    signal (SIGCHLD, SIG_IGN);
+  pid = fork ();
+  if (pid == 0)
+    {
+      FILE *fp;
+
+      if (stdin_file != NULL)
+	{
+	  fp = fopen (stdin_file, "r");
+	  if (fp != NULL)
+	    {
+	      dup2 (fileno (fp), 0);
+	      fclose (fp);
+	    }
+	}
+      if (stdout_file != NULL)
+	{
+	  unlink (stdout_file);
+	  fp = fopen (stdout_file, "w");
+	  if (fp != NULL)
+	    {
+	      dup2 (fileno (fp), 1);
+	      fclose (fp);
+	    }
+	}
+      if (stderr_file != NULL)
+	{
+	  unlink (stderr_file);
+	  fp = fopen (stderr_file, "w");
+	  if (fp != NULL)
+	    {
+	      dup2 (fileno (fp), 2);
+	      fclose (fp);
+	    }
+	}
+
+      rc = execv ((const char *) argv[0], (char *const *) argv);
+      exit (1);
```

Comment:
convention of return values is to return 0 for success and non-zero for failure. 
please confirm every return value.

Reply:
I also hope to make the files fully indented:
* c/h files by `indent`
* cpp/hpp files by `astyle`.

Reply:
@eseokoh 

the "exec" family functions return to the previous program and execute the remaining lines (exit(1) in our case) only if a runtime error has occurred on the executed program.

from the manual:
RETURN VALUE
       The exec() functions return only if an error has occurred.  The return value is -1, and errno is set to indicate the error.

How about modifying the code to:
(void) execv(...)
assert(false);
exit(1); or exit(errno);
?

Reply:
You may return `rc` for the case.
Can you also confirm other return values?

Reply:
yes, done